packages/muelu/src/Transfers/SemiCoarsen/MueLu_SemiCoarsenPFactory_kokkos_def.hpp

```diff
@@ -0,0 +1,595 @@
+    using LOView2D = Kokkos::View<LO**, DeviceType>;
+
+    // Construct a map from fine level column to layer ids (including ghost nodes)
+    // Note: this is needed to sum all couplings within a layer
+    const auto FCol2LayerVector = Xpetra::VectorFactory<LO, LO, GO, NO>::Build(Amat->getColMap());
+    const auto localTemp  = Xpetra::VectorFactory<LO, LO, GO, NO>::Build(Amat->getDomainMap());
+    RCP<const Import> importer = Amat->getCrsGraph()->getImporter();
+    if (importer == Teuchos::null)
+      importer = ImportFactory::Build(Amat->getDomainMap(), Amat->getColMap());
+    {
+      // Fill local temp with layer ids and fill ghost nodes
+      const auto localTempHost = localTemp->getHostLocalView();
+      for (int row = 0; row < NFRows; row++)
+        localTempHost(row, 0) = LayerId[row / DofsPerNode];
+      const auto localTempView = localTemp->getDeviceLocalView();
+      // const auto localTempView = localTemp->template getLocalView<DeviceType>();
+      Kokkos::deep_copy(localTempView, localTempHost);
+      FCol2LayerVector->doImport(*localTemp, *(importer), Xpetra::INSERT);
+    }
+    const auto FCol2LayerView = FCol2LayerVector->getDeviceLocalView();
+    // const auto FCol2LayerView = FCol2LayerVector->template getLocalView<DeviceType>();
+    const auto FCol2Layer = Kokkos::subview(FCol2LayerView, Kokkos::ALL(), 0);
+
+    // Construct a map from fine level column to local dof per node id (including ghost nodes)
+    // Note: this is needed to sum all couplings within a layer
+    const auto FCol2DofVector = Xpetra::VectorFactory<LO, LO, GO, NO>::Build(Amat->getColMap());
+    {
+      // Fill local temp with local dof per node ids and fill ghost nodes
+      const auto localTempHost = localTemp->getHostLocalView();
+      for (int row = 0; row < NFRows; row++)
+        localTempHost(row, 0) = row % DofsPerNode;
+      const auto localTempView = localTemp->getDeviceLocalView();
+      // const auto localTempView = localTemp->template getLocalView<DeviceType>();
+      Kokkos::deep_copy(localTempView, localTempHost);
+      FCol2DofVector->doImport(*localTemp, *(importer), Xpetra::INSERT);
+    }
+    const auto FCol2DofView = FCol2DofVector->getDeviceLocalView();
+    // const auto FCol2DofView = FCol2DofVector->template getLocalView<DeviceType>();
+    const auto FCol2Dof = Kokkos::subview(FCol2DofView, Kokkos::ALL(), 0);
+
+    // Compute NVertLines
+    // TODO: Read this from line detection factory
+    int NVertLines = -1;
+    if (NFNodes != 0) NVertLines = VertLineId[0];
+    for (int node = 1; node < NFNodes; ++node)
+      if (VertLineId[node] > NVertLines) NVertLines = VertLineId[node];
+    NVertLines++;
+
+    // Construct a map from Line, Layer ids to fine level node
+    LOView2D LineLayer2Node("LineLayer2Node", NVertLines, NFLayers);
+    typename LOView2D::HostMirror LineLayer2NodeHost = Kokkos::create_mirror_view(LineLayer2Node);
+    for (int node = 0; node < NFNodes; ++node)
```

Comment:
Why not run on device, this loops over all fine nodes on mesh

Reply:
`VertLineId` and `LayerId` are `ArrayRCP` and are constructed in LineDetectionFactory. I think I would have to refactor LineDetectionFactory to construct views instead in order to get this on device. I think I could do that but it would take more time. Let me know if you think I should go ahead with that. I was on the fence about it since I'm not sure if it would contribute much.

I suppose an alternative would be to copy them into views before the calling BuildSemiCoarsenP. Maybe that's a good middle ground.